src/Miki.Framework/Extension/ServiceExtensions.cs

```diff
@@ -0,0 +1,24 @@
+﻿﻿using System;
+using Microsoft.Extensions.DependencyInjection;
+ using Miki.Cache;
+ using Miki.Discord.Common;
+using Miki.Framework.Hosting;
+
+namespace Miki.Framework
+{
+    public static class ServiceCollectionExtensions
+    {
+        public static T GetOrCreateService<T>(this IServiceProvider provider)
+        {
+            return provider.GetService<T>() ?? ActivatorUtilities.CreateInstance<T>(provider);
+        }
+        
+        public static IServiceCollection AddCacheClient<T>(this IServiceCollection services)
+            where T : class, IExtendedCacheClient
+        {
+            services.AddSingleton<ICacheClient>(provider => provider.GetService<IExtendedCacheClient>());
```

Comment:
How does this work?

Reply:
`IExtendedCacheClient` extends `ICacheClient`. This registration redirects `ICacheClient` to `IExtendedCacheClient`, you could register it two times but that means you have (for example) two Redis connections open.

Reply:
That doesn't make sense, we already have connection pooling.